sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,680 @@
+[SYCL\_EXT\_ONEAPI\_PROPERTY\_LIST][1], which adds a general mechanism for
+specifying properties which are known at compile time.  This extension is not
+itself a feature, but rather a building block that can be incorporated into
+other features.
+
+[1]: <extensions/PropertyList/SYCL_EXT_ONEAPI_property_list.asciidoc>
+
+There are a number of situations where we plan to use compile-time constant
+properties, but this design document does not attempt to address them all.
+Rather, it describes the design for each "category" of use and illustrates each
+category with a specific feature.  For example `accessor` is used to illustrate
+properties that are applied to a kernel argument, but the same technique could
+be used for other variables that are captured as kernel arguments.
+
+In all cases, the goal of this design is to explain how a DPC++ program that
+uses properties is consumed by the device compiler and eventually represented
+in LLVM IR.  This typically involves some logic in the header files which
+results in a C++ annotation that contains the properties.  The device compiler
+front-end is responsible for consuming this annotation and producing some
+corresponding LLVM IR.  One of the goals of this design is to avoid changes to
+the front-end each time we add a new property, so the front-end is not required
+to understand each property it consumes.  Instead, it follows a mechanical
+process for converting properties listed in the C++ annotation into LLVM IR,
+and this mechanical process need not be updated when we add new properties.
+
+Once the information about properties is represented in IR, it is available to
+compiler passes.  For example, the `sycl-post-link` tool might use a property
+in order to perform one of its transformations.  Some properties are consumed
+by the DPC++ compiler, but others are transformed into SPIR-V for use by the
+JIT compiler.  This design document also describes how this SPIR-V
+transformation is done.
+
+
+## Properties on a global variable type
+
+One use for compile-time properties is with types that are used exclusively
+for declaring global variables.  One such example is the
+[SYCL\_EXT\_ONEAPI\_DEVICE\_GLOBAL][2] extension:
+
+[2]: <extensions/DeviceGlobal/SYCL_INTEL_device_global.asciidoc>
+
+```
+namespace sycl::ext::oneapi {
+
+template <typename T, typename PropertyListT = property_list<>>
+class device_global {/*...*/};
+
+} // namespace sycl::ext::oneapi
+```
+
+The following code illustrates a `device_global` variable that is declared with
+two compile-time properties:
+
+```
+using sycl::ext::oneapi;
+
+device_global<int,
+  property_list_t<
+    device_image_scope::value_t,
+    host_access::value_t<host_access::access::read>>>
+  dm1;
+```
+
+The header file represents these properties with an internal C++ attribute
+named `[[__sycl_detail__::add_ir_global_variable_attributes()]]` whose value
+is a list that is created through a template parameter pack expansion:
```

Comment:
I wanted to point this out to @elizabethandrews and @premanandrao.  Note that this design calls for using a parameter pack expansion in a C++ template.  My understanding from @erichkeane is that this is a C++ feature that we do not currently support, but it is something we want to support anyways.  Therefore, the proposed design relies on that support being implemented.

I'm not sure how much effort this is to implement.  If it is substantial, we could consider alternatives.  However, I suspect that other alternative designs will end up being more work in the front-end in the long term because we will have a continuing need for front-end changes as more properties are added.  Using the C++ attribute (with pack expansion) should help solve this.

Reply:
I have a seemingly working prototype for a C++ attribute that handles parameter packs. It seems to be doable with a minor change to the parser and then handing of parameter packs in the template instantiation of the corresponding attributes. We can even go as far as to make the parameter pack support in the parser be only for the `__sycl_detail__` namespace if we want it to seem less disruptive. So I think it is feasible.

Reply:
A more up-to-date variant is up for review upstream: https://reviews.llvm.org/D114439. I am currently adjusting for comments.